### PR TITLE
Proofread filesystems chapter

### PR DESCRIPTION
**Pull request by PradyumnaShome**
**Is this a quick PR -- Typo, Quick Wording Fix, etc? If so, you don't need to follow the template -- please delete all and make the title of the PR representative of the changes and add a member of staff to review the PR. Thank you so much!**

## Description

*Replace this paragraph with a description of what this PR is doing. If you're modifying existing behavior, describe the existing behavior, how this PR is changing it, and what motivated the change.*

## Related Issues

Addresses #82.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I updated AUTHORS with my name and email.
- [x] I read the Contributor Guide and followed the process outlined there for submitting PRs.
- [ ] My PR builds for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] I updated/added relevant documentation.
- [x] The title of the PR is descriptive and doesn't have [WIP]
- [ ] I updated CHANGELOG.md to add a description of the change.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does this cause any new material not previously covered in the course to be required?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md).
- [ ] No, this is *not* a breaking change.

### CODE REVIEW

File: filesystems/filesystems.tex [hunk 1]
```diff
@@ -296,7 +309,7 @@ \subsection{Directory API}
 Any file descriptors opened, and any memory allocated, by opendir are never released.
 This means eventually the process will run out of resources and an \keyword{open} or \keyword{opendir} call will fail.
 
-The fix is to ensure we free up resources in every possible code-path.
+The fix is to ensure we free up resources in every possible code-path (often why kernel error handlers use \keyword{goto}s extensively).
```
Comment (bhuvy2) [inline filesystems/filesystems.tex]: The real reason for gotos is not to add an additional stack frame. One could write a series of functions and make sure that resources are cleaned up at the end through nesting.

File: filesystems/filesystems.tex [hunk 2]
```diff
@@ -354,108 +367,109 @@ \subsection{Directory API}
...
-
-However, symlinks have a key disadvantage, they as slower than regular files and directories.
-When the links contents are read, they must be interpreted as a new path to the target file, resulting in an additional call to open and read since the real file must be opened and read.
-Another disadvantage is that POSIX will not let you hard link directories whereas soft links are allowed.
-The \keyword{ln} command will only allow root to do this and only if you provide the \keyword{-d} option.
-However, even root may not be able to perform this because most filesystems prevent it!
+\begin{enumerate}
+    \item \keyword{Hard Links}
+        A hard link is simply an entry in a directory assigning some name to an inode number that already has a different name and mapping in either the same directory or a different one.
+        If we already have a file on a file system we can create another link to the same inode using the `ln' command:
```
Comment (bhuvy2) [inline filesystems/filesystems.tex]: `\keyword{ln}`

File: filesystems/filesystems.tex [hunk 3]
```diff
@@ -354,108 +367,109 @@ \subsection{Directory API}
...
+
+        \begin{verbatim}
+        $ ln file1.txt blip.txt
+        \end{verbatim}
```
Comment (bhuvy2) [inline filesystems/filesystems.tex]: We should change these to `\begin{lstlisting}`

File: filesystems/filesystems.tex [hunk 4]
```diff
@@ -354,108 +367,109 @@ \subsection{Directory API}
...
+        Hard links can be created anywhere inside the same filesystem.
+
+    \item \keyword{Soft Links}
+        The second kind of link is a soft link - or a symbolic link or a symlink.
+        A symbolic link is different because it does not deal with inode numbers directly.
+        Instead, a symbolic link is a file on disk with a special bit set and stores a path to another file.
+        Quite simply, without the special bit, it is nothing more than a text file with a file path inside.
+        Note when people generally talk about a link without specifying hard or soft, they are referring to a hard link.
+
+        To create a symbolic link in the shell use \keyword{ln -s}.
```
Comment (bhuvy2) [inline filesystems/filesystems.tex]: `in the shell, use`

File: filesystems/filesystems.tex [hunk 5]
```diff
@@ -354,108 +367,109 @@ \subsection{Directory API}
...
+        To read the contents of the link as just a file use \keyword{readlink}.
```
Comment (bhuvy2) [inline filesystems/filesystems.tex]: as just a file, use

File: filesystems/filesystems.tex [hunk 6]
```diff
@@ -467,18 +481,20 @@ \subsection{Linking}
...
-However, there are some nuances. It is possible to have a path called \texttt{a/b/../c/./}.
+Now that we have definitions, and have talked about directories, we come across the concept of a path.
+A path is a sequence of directories that provide one with a "path" in the graph that is a filesystem.
+However, there are some nuances.
+It is possible to have a path called \texttt{a/b/../c/./}.
 Since \keyword{..} and \keyword{.} are special entries in directories, this is a valid path that actually refers to \texttt{a/c}.
 Most filesystem functions will allow uncompressed paths to be passed in.
-The C library provides a function \keyword{realpath} to compress the path or get the realpath.
+The C library provides a function \keyword{realpath} to compress the path or get the absolute path.
 To simplify by hand remember that \keyword{..} means `parent folder' and that \keyword{.} means `current folder'.
```
Comment (bhuvy2) [inline filesystems/filesystems.tex]: To simplify by hand, remember

File: filesystems/filesystems.tex [hunk 7]
```diff
@@ -517,23 +533,25 @@ \subsection{Metadata}
...
-FILE *file = fopen("notes.txt", "r");
-int fd = fileno(file); /* Just for fun - extract the file descriptor from a C FILE struct */
-struct stat s;
-fstat(fd, & s);
-printf("Last accessed %s", ctime(&s.st_atime));
+    FILE *file = fopen("notes.txt", "r");
+    int fd = fileno(file); /* Just for fun - extract the file descriptor from a C FILE struct */
+    struct stat s;
+    fstat(fd, & s);
+    printf("Last accessed %s", ctime(&s.st_atime));
```
Comment (bhuvy2) [inline filesystems/filesystems.tex]: When we put in spacing to a lstlisting environment, it creates an odd indent in the markdown and the pdf.

File: filesystems/filesystems.tex [hunk 8]
```diff
@@ -592,40 +611,40 @@ \section{Permissions and bits}
...
 Execute will allow you to enter the directory and access it.
-Without the execute bit it is not possible to create or remove files or directories since you cannot access them.
+Without the execute bit, it is not possible to create or remove files or directories since you cannot access them.
 You can, however, list the contents of the directory.
 
 There are several command line utilities for interacting with a file's mode.
 \keyword{mknod} changes the type of the file.
 \keyword{chmod} takes a number and a file and changes the permission bits.
-However, before we can discuss chmod in detail, we must also understand the user id (\keyword{uid}) and group id (\keyword{gid}) as well.
+However, before we can discuss chmod in detail, we must also understand the user ID (\keyword{uid}) and group id (\keyword{gid}) as well.
```
Comment (bhuvy2) [inline filesystems/filesystems.tex]: However before we can discuss ...

File: filesystems/filesystems.tex [hunk 9]
```diff
@@ -738,25 +757,25 @@ \subsection{The `sticky' bit}
...
+    drwxr-xr-x 53 aneesh aneesh    4096 Nov  1 14:19 ..
+    drwxr-xr-t  2 aneesh aneesh    4096 Nov  1 14:19 sticky
+    aneesh$ su newuser
+    newuser$ rm -rf sticky
+    rm: cannot remove 'sticky': Permission denied
+    newuser$ exit
+    aneesh$ rm -rf sticky
+    aneesh$ ls -l
+    drwxr-xr-x  7 aneesh aneesh    4096 Nov  1 14:19 .
+    drwxr-xr-x 53 aneesh aneesh    4096 Nov  1 14:19 ..
```
Comment (bhuvy2) [inline filesystems/filesystems.tex]: Let's keep the spaces out of verbatim too.

This should also be an lstlisting.

File: filesystems/filesystems.tex [hunk 10]
```diff
@@ -766,19 +785,18 @@ \section{Virtual filesystems and other filesystems}
...
+Linux provides 3 main virtual filesystems:
 
 \begin{verbatim}
-/dev  - A list of physical and virtual devices (for example network card, cdrom, random number generator)
-/proc - A list of resources used by each process and (by tradition) set of system information
-/sys - An organized list of internal kernel entities
+    /dev  - A list of physical and virtual devices (for example network card, cdrom, random number generator)
+    /proc - A list of resources used by each process and (by tradition) set of system information
+    /sys - An organized list of internal kernel entities
 \end{verbatim}
```
Comment (bhuvy2) [inline filesystems/filesystems.tex]: This should probably be a list or table.

File: filesystems/filesystems.tex [hunk 11]
```diff
@@ -810,7 +828,7 @@ \subsection{Managing files and filesystems}
 A better version is to atomically create the directory with the correct permissions from its inception.
 
 \begin{verbatim}
-$ mkdir -m 700 /tmp/mystuff
+    $ mkdir -m 700 /tmp/mystuff
```
Comment (bhuvy2) [inline filesystems/filesystems.tex]: Indent here too.

File: filesystems/filesystems.tex [hunk 12]
```diff
@@ -1104,34 +1122,38 @@ \subsection{Higher Level Raids}
...
-Finally, we have an array of inodes followed by the rest of the disk - implicitly partitioned into data blocks. It's
-important to note that there may not be any real distinction from one data block to the next from the perspective of the
-hardware components of the disk. Thinking about the disk as an array of data blocks is simply something we do so that we
-have a way to describe where files live on disk.
-
-Below, we have an example of how an inode that describes a file may look. Note that for the sake of simplicity, we have
-drawn arrows mapping data block numbers in the inode to their locations on disk, these aren't pointers so much as
-indices into an array.
+Software developers need to implement filesystems all the time.
+Don't believe me? Take a look at Hadoop, GlusterFS, Qumulo, etc.
```
Comment (bhuvy2) [inline filesystems/filesystems.tex]: Maybe we should take out the Don't believe me part or restructure sentence.

File: filesystems/filesystems.tex [hunk 13]
```diff
@@ -1104,34 +1122,38 @@ \subsection{Higher Level Raids}
...
+Filesystems are hot areas of research at the time of writing, because people have realized that the software models that we have devised don't take full advantage of our current hardware.
+Additionally, the hardware that we use for storing information is getting better all the time.
+As such, you may end up designing a filesystem yourself someday.
+In this section we will go over one of our fake filesystems that we talk about in class and ``walk through'' some examples of how things work.
```
Comment (bhuvy2) [inline filesystems/filesystems.tex]: In this section, we

File: filesystems/filesystems.tex [hunk 14]
```diff
@@ -1104,34 +1122,38 @@ \subsection{Higher Level Raids}
...
+
+So what does our hypothetical filesystem look like?
```
Comment (bhuvy2) [inline filesystems/filesystems.tex]: So, what

File: filesystems/filesystems.tex [hunk 15]
```diff
@@ -1159,82 +1181,100 @@ \subsection{How big is our file?}
...
+    \item What about when the first indirect entry (\#4) is full?
+\end{itemize}
+
+These are edge cases you have to think about for a filesystem that really get you.
 
 \subsection{Adding Deletes}
 
-Although this isn't part of the lab originally.
-If you were to ask what happens when a file gets deleted, it's actually pretty simple.
+Although this wasn't part of the CS 241 assignment originally, if you were to ask what happens when a file gets deleted, it is actually quite simple.
```
Comment (bhuvy2) [inline filesystems/filesystems.tex]: We could just get rid of this sentence. The less tie in we have with our assignments, the better.

File: filesystems/filesystems.tex (hunk 1 above)
Comment (PradyumnaShome) [reply filesystems/filesystems.tex]: Okay, I'll remove it then.

File: filesystems/filesystems.tex (hunk 10 above)
Comment (PradyumnaShome) [reply filesystems/filesystems.tex]: I thought so too, but I wasn't sure.

**Pull request merged**